File: src/main/kotlin/uk/ac/ic/matterialize/matting/backgroundcut/models/PixelBgModel.kt
package uk.ac.ic.matterialize.matting.backgroundcut

import org.opencv.core.CvType
import org.opencv.core.Mat
import org.opencv.core.Size
import kotlin.math.exp

class PixelBgModel(bgImage: Image) {
    private val perPixelVariances = bgImage.getPixelVariances()
    private val flatBgImage = bgImage.flattened

    fun pdf(l2: Double, det: Double): Double {
        val x = -1 * l2 / 2
        return exp(x) / kotlin.math.sqrt(twoPiCubed * det)
    }

    fun perPixelProbs(image: FlatImage): Probs {
        val probs = Mat(Size(1.0, image.numPixels().toDouble()), CvType.CV_64FC1)

        for (pixel in 0 until image.numPixels()) {
            var l2 = 0.0
            var det = 1.0
            for (channel in 0 until 3) {
                val delta = image.get(pixel, channel).toDouble() - (flatBgImage.get(pixel, channel)).toDouble()
                l2 += (delta * delta) / perPixelVariances.at(pixel, channel)
                det *= perPixelVariances.at(pixel, channel)
            }

            probs.put(pixel, 0, pdf(l2, det))
        }

        return Probs(probs)
    }
}

File: src/matting/background_cut/models/pixel_bg_model.cpp
#include "pixel_bg_model.hpp"

#include <cmath>
#include <utility>

#include <opencv2/core.hpp>

constexpr auto PI{3.14159265358979323846};

double PixelBgModel::pdf(double l2, double det) const {
  return std::exp(-(1 * l2) / 2) / std::sqrt(8 * PI * PI * PI * det);
}

Probability PixelBgModel::get_pixel_probability(const Image &image) const {
  cv::Mat probabilities(
      cv::Size2d(1.0, static_cast<double>(image.num_pixels())), CV_64FC1);

  for (int pixel{0}; pixel < image.num_pixels(); ++pixel) {
    double l2{0};
    double det{1.0};

    for (int channel{}; channel < 3; ++channel) {
      const double delta{
          static_cast<double>(image.get(pixel, channel)) -
          static_cast<double>(bg_image.get(pixel, channel))};
      const double variance{per_pixel_variance.at(pixel, channel)};
      // delta / variance can cause rounding issues to 0, hence brackets
      // required for safety with high variance values
      l2 += (delta * delta) / variance;
      det *= variance;
    }

    *probabilities.ptr<double>(pixel) =
        std::min<double>(pdf(l2, det) * magic_scaling, 1);
  }

  return Probability(std::move(probabilities));
}
